File: packages/mirascript/src/vm/lib/global/math-arr.ts
import type { VmAny } from '../../types/index.ts';
import { getNumbers, VmLib } from '../_helpers.ts';

/** 生成函数 */
function build(f: (...values: readonly number[]) => number): (...values: readonly VmAny[]) => number {
    return (...values) => {
        const numbers = getNumbers(values);
        return f(...numbers);
    };
}

export const max = VmLib(build(Math.max), {
    summary: '返回一组数中的最大值',
    params: { '..values': '要比较的数值' },
    paramsType: { '..values': 'number[]' },
    returnsType: 'number',
    examples: ['max(3, 7, 2) // 7'],
});

export const min = VmLib(build(Math.min), {
    summary: '返回一组数中的最小值',
    params: { '..values': '要比较的数值' },
    paramsType: { '..values': 'number[]' },
    returnsType: 'number',
    examples: ['min(3, 7, 2) // 2'],
});

export const hypot = VmLib(build(Math.hypot), {
    summary: '返回所有参数平方和的平方根',
    params: { '..values': '要计算的数值' },
    paramsType: { '..values': 'number[]' },
    returnsType: 'number',
    examples: ['hypot(3, 4) // 5'],
});

export const sum = VmLib(
    (...values: readonly VmAny[]) => {
        const numbers = getNumbers(values);
        return numbers.reduce((a, b) => a + b, 0);
    },
    {
        summary: '返回一组数的总和',
        params: { '..values': '要计算的数值' },
        paramsType: { '..values': 'number[]' },
        returnsType: 'number',
        examples: ['sum(1, 2, 3, 4) // 10'],
    },
);

export const product = VmLib(
    (...values: readonly VmAny[]) => {
        const numbers = getNumbers(values);
        return numbers.reduce((a, b) => a * b, 1);
    },
    {
        summary: '返回一组数的乘积',
        params: { '..values': '要计算的数值' },
        paramsType: { '..values': 'number[]' },
        returnsType: 'number',
        examples: ['product(2, 3, 4) // 24'],
    },
);

File: crates/python/mirascript/vm/lib/vm_global/math_arr.py
from .._helpers import get_numbers
from functools import reduce as _reduce
import math
import sys
from ...types.checker import is_positive_zero, is_negative_zero


def build(func):
    def wrapper(*args):
        numbers = get_numbers(args)
        return func(*numbers)

    return wrapper


def _max(*args):
    high = -math.inf
    for num in args:
        if math.isnan(num):
            return math.nan
        if is_positive_zero(num) and is_negative_zero(high):
            high = num
        elif num > high:
            high = num
    return high


max_ = build(_max)


def _min(*args):
    low = math.inf
    for num in args:
        if math.isnan(num):
            return math.nan
        if is_negative_zero(num) and is_positive_zero(low):
            low = num
        elif num < low:
            low = num
    return low


min_ = build(_min)


def hypot(*args):
    numbers = get_numbers(args)
    if len(numbers) == 0:
        return 0.0
    # math.hypot accepts an arbitrary number of args starting Python 3.8.
    # For older Pythons, provide a compatible fallback.
    if sys.version_info >= (3, 8):
        return math.hypot(*numbers)

    if len(numbers) == 1:
        return abs(numbers[0])
    if len(numbers) == 2:
        return math.hypot(numbers[0], numbers[1])

    # Manual computation for >2 args on older Pythons
    s = 0.0
    for n in numbers:
        if math.isnan(n):
            return math.nan
        s += n * n
    return math.sqrt(s)


sum_ = build(lambda *args: _reduce(lambda a, b: a + b, args, -0.0))
product = build(lambda *args: _reduce(lambda a, b: a * b, args, 1.0))
